tests/integration_tests/functional/test_balloon.py

```diff
@@ -45,22 +45,28 @@ def get_rss_from_pmap():
 
 def make_guest_dirty_memory(ssh_connection, should_oom=False, amount=8192):
     """Tell the guest, over ssh, to dirty `amount` pages of memory."""
+    logger = logging.getLogger("make_guest_dirty_memory")
+
     amount_in_mbytes = amount / MB_TO_PAGES
 
-    exit_code, _, _ = ssh_connection.execute_command(
-        "/sbin/fillmem {}".format(amount_in_mbytes)
-    )
+    cmd = f"/sbin/fillmem {amount_in_mbytes}"
+    exit_code, stdout, stderr = ssh_connection.execute_command(cmd)
+    # add something to the logs for troubleshooting
+    if exit_code != 0:
+        logger.error("while running: %s", cmd)
+        logger.error("stdout: %s", stdout.read())
+        logger.error("stderr: %s", stderr.read())
 
     cmd = "cat /tmp/fillmem_output.txt"
     _, stdout, _ = ssh_connection.execute_command(cmd)
     if should_oom:
         assert (
-            exit_code == 0
-            and ("OOM Killer stopped the program with " "signal 9, exit code 0")
-            in stdout.read()
+            "OOM Killer stopped the program with "
+            "signal 9, exit code 0" in stdout.read()
         )
     else:
-        assert exit_code == 0 and ("Memory filling was " "successful") in stdout.read()
+        assert exit_code == 0, stderr.read()
+        assert "Memory filling was successful" in stdout.read(), stdout.read()
```

Comment:
Does this double call to `.read()` return the same data twice? I would expect it not too, but I am also not familiar with the python API in question

Reply:
Yes, it returns same data

Reply:
Good catch! I should have checked if there are subsequent lines. Since the file `/tmp/fillmem_output.txt` seems to have only a single line, the second one seems to return an empty line.
https://github.com/firecracker-microvm/firecracker/blob/main/resources/tests/fillmem.c#L62-L73

Reply:
> Yes, it returns same data

Oh, Really? As `stdout` looks `StringIO` type, it doesn't return the same data in my understanding...

https://github.com/firecracker-microvm/firecracker/blob/2a5a6bc7155959d73f76f2af15125b7fd2798013/tests/host_tools/network.py#L40-L43
```
    def execute_command(self, cmd_string):
        """Execute the command passed as a string in the ssh context."""
        exit_code, stdout, stderr = self._exec(cmd_string)
        return exit_code, StringIO(stdout), StringIO(stderr)
```

E.g.,
```
>>> from io import StringIO
>>> s = StringIO("hello")
>>> assert "world" in s.read(), s.read()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError
```

Reply:
You are right @zulinx86 , I remember trying that out and working. My bad, sorry

Reply:
Good catch! I didn't notice either! fixed in last version.